overlord/ifacestate/handlers.go

```diff
@@ -989,7 +994,45 @@ func batchConnectTasks(st *state.State, snapsup *snapstate.SnapSetup, conns map[
 	if len(ts.Tasks()) > 0 {
 		ts.AddTask(setupProfiles)
 	}
-	return ts, nil
+	return ts, hasHooks, nil
+}
+
+// firstTaskAfterBootWhenPreseeding finds the first task to be run for thisSnap
+// on first boot after mark-preseeded task, this is normally install hook or
+// start-snap-services.
+func firstTaskAfterBootWhenPreseeding(thisSnap string, markPreseeded *state.Task) (*state.Task, error) {
+	var firstTaskAfterBoot *state.Task
+	if markPreseeded.Change() == nil {
+		return nil, fmt.Errorf("internal error: %s task not in change", markPreseeded.Kind())
+	}
+	for _, ht := range markPreseeded.Change().Tasks() {
```

Comment:
couldn't this just go over markPreseed.HatlTasks ?

Reply:
It could, but then it needs to check halt tasks recursively and becomes slightly more complex, not sure it is beneficial to use HaltTasks?

Reply:
I'm not understanding why it would need to be recursive, aren't the tasks we are interested in immediately after markPreseeded by definition?

Reply:
Ah sorry, you're right, I got confused by artificial scenario in my test where I had a fake "some-task" task between them.

Reply:
Updated. Also simplified it because install hook is always present, there is no need to look for start-snap-services